compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -7365,18 +7365,25 @@ private void reorderArguments(BLangInvocation iExpr) {
                 return;
             }
 
+            BArrayType restParamArrayType = (BArrayType) invokableSymbol.restParam.type;
+            BType elemType = restParamArrayType.eType;
+
             BLangArrayLiteral arrayLiteral = (BLangArrayLiteral) TreeBuilder.createArrayLiteralExpressionNode();
+            arrayLiteral.type = restParamArrayType;
             List<BLangExpression> exprs = new ArrayList<>();
-
-            BArrayType arrayType = (BArrayType) invokableSymbol.restParam.type;
-            BType elemType = arrayType.eType;
-
             for (BLangExpression restArg : restArgs) {
-                exprs.add(addConversionExprIfRequired(restArg, elemType));
+                // If the restArg is an array (where expected elemType is not an array),
+                // we have to flatten that and create a single array.
+                if (elemType.tag != TypeTags.ARRAY && restArg.getKind() == NodeKind.ARRAY_LITERAL_EXPR) {
```

Comment:
According to the comment in L7362, this should not happen, right? 

If it is a list that is passed for the rest param, the node kind should have been REST_ARGS_EXPR here. 

I think we should look into if there is an issue in how we're desugaring this, the order, or if we are desugaring something that has already been desugared.

Reply:
Correct, but we are creating the array literal for the rest args here, and when the invocation is chained, that array literal will get passed to the next chained invocation, and L7375 cast tries to cast it to the rest param type. i.e `<string> [foo, foo1]`. So it's not `REST_ARGS_EXPR`, and it'll be the array literal we create in the previous iteration and that's why it comes here.

Reply:
This happens because we're desugaring the expression on which the method is called twice, right? 

Shouldn't both the following be desugared in the same way.

```ballerina
int e = 0;
int a = e.max(1).min(2);
```

```ballerina
int e = 0;
int a = int:min(int:max(e, 1), 2);
```

Since the latter works and similar object method calls work, this could specifically be an issue due to how we add the expression on which the method is called as the first arg for langlib methods.

I guess we may have to revisit that logic anyway when fixing https://github.com/ballerina-platform/ballerina-lang/issues/28580.

Reply:
@MaryamZi, updated the PR 666f225.